docs/swift/DesignDoc.md

```diff
@@ -0,0 +1,674 @@
+# Protocol Buffers in Swift
+
+## Objective
+
+This document describes the user-facing API and internal implementation of
+proto2 and proto3 messages in Apple’s Swift programming language.
+
+One of the key goals of protobufs is to provide idiomatic APIs for each
+language. In that vein, **interoperability with Objective-C is a non-goal of
+this proposal.** Protobuf users who need to pass messages between Objective-C
+and Swift code in the same application should use the existing Objective-C proto
+library. The goal of the effort described here is to provide an API for protobuf
+messages that uses features specific to Swift—optional types, algebraic
+enumerated types, value types, and so forth—in a natural way that will delight,
+rather than surprise, users of the language.
+
+## Naming
+
+*   By convention, both typical protobuf message names and Swift structs/classes
+    are `UpperCamelCase`, so for most messages, the name of a message can be the
+    same as the name of its generated type. (However, see the discussion below
+    about prefixes under [Packages](#packages).)
+
+*   Enum cases in protobufs typically are `UPPERCASE_WITH_UNDERSCORES`, whereas
+    in Swift they are `lowerCamelCase` (as of the Swift 3 API design
+    guidelines). We will transform the names to match Swift convention, using
+    a whitelist similar to the Objective-C compiler plugin to handle commonly
+    used acronyms.
```

Comment:
There's some discussion about trimming prefixes for one of the other languages.  We should dig it up and see where it's headed also.

While I agree this ends up nice for Swift, the catch is a lot of times developers end up having to work with their protos in >1 language, so there are times where consistency ends up helping also.

Oh, just remembered the problem cases that came up in the other thread:

```
enum Foo {
  FOO_SOMETHING = 0;
  FOO_ELSE = 1;
  SOMETHING_WITHOUT_FOO = 2;
}
```
Sadly, this does happen non trivial number of times.

And even worse:

```
enum Foo {
  DEFAULT = 1;
  FOO_ELSE = 1;
  FOO_DEFAULT = 2;
}
```

Yes, this is confusing, but usually there a comment to explain what happened and what the differences are.

In both cases, the problem can be proto evolution.  The proto might initially get shipped where stripping the prefix works just fine, but then a new value gets added, and it hits one of these cases; if a new generation would cause new names to happen, it means it breaks code.  And since not all languages do prefix stripping, other code might already be using the new additions so "fixing" the updated proto is also a breaking event for folks.